src/main/xerus/monstercat/api/Playlist.kt

```diff
@@ -0,0 +1,87 @@
+package xerus.monstercat.api
+
+import javafx.beans.property.SimpleBooleanProperty
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import mu.KotlinLogging
+import xerus.ktutil.javafx.properties.SimpleObservable
+import xerus.ktutil.javafx.properties.bindSoft
+import xerus.monstercat.api.response.Track
+import java.util.*
+
+object Playlist {
+	val logger = KotlinLogging.logger { }
+	
+	val tracks: ObservableList<Track> = FXCollections.observableArrayList()
+	val history = ArrayDeque<Track>()
+	val currentIndex = SimpleObservable<Int?>(null).apply {
+		bindSoft({
+			tracks.indexOf(Player.activeTrack.value).takeUnless { i -> i == -1 }
+		}, Player.activeTrack, tracks)
+	}
+	
+	val repeat = SimpleBooleanProperty(false)
+	val shuffle = SimpleBooleanProperty(false)
+	
+	operator fun get(index: Int): Track? = tracks.getOrNull(index)
+	
+	/** This property prevents songs that come from the [history] from being added to it again,
+	 * which would effectively create an infinite loop */
+	var lastPolled: Track? = null
+	
+	fun getPrev(): Track? = history.pollLast()?.also {
+		lastPolled = it
+		logger.debug("Polled $it from History")
+	} ?: currentIndex.value?.let { get(it - 1) }
+	
+	fun getNext() = when {
+		shuffle.value -> getNextTrackRandom()
+		repeat.value && (getNextTrack() == null) -> tracks[0]
+		else -> getNextTrack()
+	}
+	
+	fun addNext(track: Track) {
+		tracks.remove(track)
+		tracks.add(currentIndex.value?.let { it + 1 } ?: 0, track)
+	}
+	
+	fun add(track: Track) {
+		tracks.remove(track)
+		tracks.add(track)
+	}
+	
+	fun removeAt(index: Int?) {
+		tracks.removeAt(index ?: tracks.size - 1)
+	}
+	
+	fun clear() {
+		history.clear()
+		tracks.clear()
+	}
+	
+	fun setTracks(playlist: Collection<Track>) {
+		history.clear()
+		tracks.setAll(playlist)
+	}
+	
+	fun getNextTrackRandom(): Track {
+		val index = (Math.random() * (tracks.size - 1)).toInt().let { if(it >= currentIndex.value!!) it + 1 else it }.takeUnless { it >= tracks.size }
```

Comment:
1. `size - 1` = `lastIndex` - in most contexts the latter is semantically the correct choice and easier to understand.
2. Instead of Math.random and then toInt you could probably use Random.randInt
3. `takeUnless { it >= tracks.size }` - how can it be larger than tracks.size?
4. Please break this line up in a way that makes it easier to understand.

Reply:
3. If there is 1 track in the playlist, it will end up being equal to tracks.size (>= 1) and won't work because array calls starts at 0

Reply:
ab8c72faa1d51cdc66c3e70e63ec63b0d52b6890

Reply:
3. If that is the only case where this can happen, you should filter this case out beforehand instead of adding an irritating clause.

Reply:
Sure.

Reply:
a062f9cb127f6148df93aec0ac16d029a3814776